stdlib/LinearAlgebra/src/eigen.jl

```diff
@@ -647,9 +647,34 @@ function Base.isequal(A::Eigen, B::Eigen)
 end
 
 # Conversion methods
-
 ## Can we determine the source/result is Real?  This is not stored in the type Eigen
-AbstractMatrix(F::Eigen) = F.vectors * Diagonal(F.values) / F.vectors
+AbstractMatrix(F::Eigen) = F.are_vectors_unitary ?
+                             (F.vectors * Diagonal(F.values) * (F.vectors')) :
+                             (F.vectors * Diagonal(F.values) / F.vectors)
 AbstractArray(F::Eigen) = AbstractMatrix(F)
 Matrix(F::Eigen) = Array(AbstractArray(F))
 Array(F::Eigen) = Matrix(F)
+
+# Operations on Eigen
+# Only operations that can be efficiently implemented or conserve the eigen type are implemented
+adjoint(E::Eigen) = Eigen(adjoint.(E.values),
+                          copy(adjoint(E.are_vectors_unitary ? (E.vectors') : inv(E.vectors))),
+                          E.are_vectors_unitary)
+transpose(E::Eigen) = Eigen(transpose.(E.values),
+                            copy(transpose(E.are_vectors_unitary ? (E.vectors') : inv(E.vectors))),
+                            E.are_vectors_unitary && (eltype(E) <: Real))
+
+det(A::Eigen) = prod(A.values)
+tr(E::Eigen) = sum(E.values)
+
+*(E::Eigen, x::AbstractVector) = E.vectors * (E.values .* (E.are_vectors_unitary ? (E.vectors'x) : (E.vectors \ x)))
+*(x::Adjoint{<:Any, <:AbstractVector}, E::Eigen) = E.are_vectors_unitary ?
+                                                   ((x*(E.vectors)) * Diagonal(E.values)) * (E.vectors') :
+                                                   ((x*(E.vectors)) * Diagonal(E.values)) / E.vectors
+mul!(x, E::Eigen, y::AbstractVector) = copyto!(x, E*y)
+
+inv(E::Eigen) = Eigen(map(inv, E.values), copy(E.vectors), E.are_vectors_unitary)
+for f in _matrix_functions
```

Comment:
The restriction of applicable functions to this list is a burden. What about adding an additional function
like `apply(f, E)` = ... to allow for user defined functions? 
remark:
unfortuantely `(f::Function)(E) = ...` is not supported by Julia.

Reply:
I'm concerned about exporting a function here, esp since the functionality is relatively half-baked (it doesn't apply to general matrices, just to Eigen). I think it would make a lot more sense to have a MatrixFunctions package that does this properly (including for non-hermitian matrices)

Reply:
> Also your argument is valid for sin, cosh, exp,  and brothers in the same way as for apply, so I don't understand your concern.

Adding methods to these functions does not add API surface

> This operation is valid for all diagonalizable matrices, not only Hermitian or Symmetric, or normal ones.

Technically, yes. In practice, using it for strongly non-normal matrices is a bad idea (see eg the classic paper "N bad ways to compute matrix exponentials" or whatever it's called), and there are better ways to do it. Ideally, a MatrixFunctions package would deal with this. This PR is just an opportunity to define methods that ought to exist already.